system/security/shared/caching.cpp

```diff
@@ -691,19 +710,24 @@ void CPermissionsCache::managedFileScopesCacheFillThread()
         removeAllManagedFileScopes();
         if (count)
             addManagedFileScopes(scopes);
-        time(&m_lastCacheFillTime);
+        time_t now;
+        time(&now);
+        m_lastCacheFillTime = now;
     }
 
-    unsigned int waitTimeSeconds = m_cacheTimeoutInSeconds;
+    unsigned waitTimeSeconds = m_cacheTimeoutInSeconds;
     while (!m_stopFileScopeCacheFillThread && m_cacheTimeoutInSeconds)
     {
         m_fileScopeCacheFillSem.wait(waitTimeSeconds * 1000);
 
         time_t now;
         time(&now);
 
+        // Save previous fill time in case another thread modifies it
+        time_t previousCacheFillTime = m_lastCacheFillTime;
```

Comment:
m_lastCacheFillTime could have just been set at this point, i.e. after line 724,
and therefore it can be > 'now'.

Meaning (now-previousCacheFillTime) can go negative.
if (now- previousCacheFillTime) is negative, then the expression on line 730 will evaluate to true - incorrectly.

As per https://github.com/hpcc-systems/HPCC-Platform/pull/19673#discussion_r2039887586, this needs to explicitly check for that, and not enter update block.